File: easy/0155minStack/minStack.js
/**
 * initialize your data structure here.
 */
const MinStack = function () {
    this.stack = [];
    this.minStack = [];
}

/** 
 * @param {number} element
 * @return {void}
 */
MinStack.prototype.push = function(element) {
    this.stack.push(element);
    if (this.minStack.length == 0 || x <= this.minStack[this.minStack.length - 1]) {
        //只能放进一个比minStack最后一项小的数
        this.minStack.push(element);
    }
}

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
    const pop = this.stack.pop();
    if (pop !== void 0 && pop === this.minStack[this.minStack.length - 1]) {
        this.minStack.pop();
    }
}

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
    return this.stack[this.stack.length - 1];
}

/**
 * @return {number}
 */
MinStack.prototype.min = function () {
    return this.minStack[this.minStack.length - 1];
}

MinStack.prototype.print = function () {
    console.log(this.minStack.toString());
}


// Your MinStack object will be instantiated and called as such:
 var obj = new MinStack()
 obj.push("x")
 obj.print()
 obj.pop()
 var param_3 = obj.top()
 var param_4 = obj.min()
 obj.print()
 

File: easy/0155minStack/minStack.java
import java.util.*;

public class minStack{

    // 数据栈
    private Stack<Integer> data;
    // 辅助栈
    private Stack<Integer> helper;

    /**
     * initialize your data structure here.
     */
    public minStack() {
        data = new Stack<Integer>();
        helper = new Stack<Integer>();
    }

    public void push(int x) {
        // 辅助栈在必要的时候才增加
        data.add(x);
        if(helper.isEmpty() || x <= helper.peek()) {
            helper.add(x);
        }
    }

    public void pop() throws NoSuchElementException {
        if(data.isEmpty()) {
            throw new NoSuchElementException("can't pop from an empty stack");
        }
        int top = data.pop();
        if(top == helper.peek()){
            helper.pop();
        }
    }

    public int top() throws NoSuchElementException {
        if(data.isEmpty()) {
            throw new NoSuchElementException("can't find existing element");
        }
        return data.peek();
    }

    public int getMin() throws NoSuchElementException {
        if(data.isEmpty()) {
            throw new NoSuchElementException("can't find existing element");
        }
        return helper.peek();
    }

    public void print() {
        System.out.println("element in data: ");
        for (int element : data) {
            System.out.println(element);
        }
        System.out.println("");
        System.out.println("element in helper: ");
        for (int element : helper) {
            System.out.println(element);
        }
    }

}